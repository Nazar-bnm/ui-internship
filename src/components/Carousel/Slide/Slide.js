import React from 'react';

import PropTypes from 'prop-types';
import './Slide.scss';

import cx from 'classnames';

export const CN = 'slide';

const Slide = ({ image, name, category }) => (
  <div className={cx(CN)}>
    <img
      className={`${CN}__image`}
      src={image}
      alt={name}
    />
    <h3 className={`${CN}__product-name title-sections`}>{name}</h3>
    <p className={`${CN}__category`}>{category}</p>
  </div>
);

Slide.propTypes = {
  image: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  category: PropTypes.string.isRequired
};

export default Slide;
