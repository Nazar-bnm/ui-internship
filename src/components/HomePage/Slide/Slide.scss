@import '@/scss/_global_styles';

.slide-info {
  width: 44%;
  padding: 1rem;
  background-color: $pale-grey;
  text-align: center;

  &__title {
    font: 0.8rem font(primary-light);
    text-transform: uppercase;
  }

  &__description {
    margin: 0.5rem 0 1.5rem;
    font: 2.5rem font(secondary);
    font-variant: small-caps;
    line-height: 1;
    text-transform: capitalize;
  }

  &__button {
    transition: transform 0.5s linear;
    border: none;
    border-radius: 2px;
    padding: 10px 15px;
    background: $black;
    cursor: pointer;
    font: 0.7rem font(primary-light);
    text-transform: uppercase;
    color: $white;

    &:hover,
    &:active {
      transform: scale(1.1);
    }
  }
}

@include media-for(tablet) {
  .slide-info {
    width: 70%;
  }
}

@include media-for(mobile-landscape) {
  .slide-info {
    width: 80%;

    &__description {
      font-size: 1.5rem;
    }
  }
}
