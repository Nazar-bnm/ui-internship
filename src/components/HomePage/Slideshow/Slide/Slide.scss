@import '../../../../scss/global_styles';

.slide {
  &-container {
    width: 100vw;
    height: 100%;    
  }

  &-wrapper {
    @include flex($direction: column, $justify: center, $align: center);
    width: 100%;
    height: 100%;
    font-family: font(primary-light);
    background-repeat: no-repeat;
    background-size: cover;
    background-position: center center;

    &-title {
      text-transform: uppercase; 
      font: 0.8rem inherit;
    }
  
    &-description {
      margin: 0.5rem 0 1.5rem; 
      font: 2.5rem font(secondary);
      line-height: 1;
      font-variant: small-caps;
      text-transform: capitalize;
    }

    &-button {
      border: none;
      padding: 5px 10px;
      font: 0.7rem inherit;
      text-transform: uppercase; 
      background: $black;
      color: $white;
    }
  }
}

.fade {
  animation-name: fade;
  animation-duration: 3s;
  opacity: 1;
}

.zoom-in {
  animation: 2s linear  forwards zoom-in;
  opacity: 1;
  transform: scale(1);

}

.zoom-out {
  animation: 2s linear  forwards zoom-out;
  opacity: 1;
  transform: scale(1);

}

@keyframes fade {
  from {opacity: 0}
  to {opacity: 1}
}

@keyframes zoom-in {
  0% {transform: scale(1); opacity: 1;}
  100% {transform: scale(0.1); opacity: 0 ;}
}


@keyframes zoom-out {
  0% {transform: scale(1); opacity: 1;}
  100% {transform: scale(1.5); opacity: 0 ;}
}

