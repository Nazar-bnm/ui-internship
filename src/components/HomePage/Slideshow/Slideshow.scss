@import '@/scss/_global_styles.scss';

.slideshow {
  position: relative;
  z-index: 1;
  overflow: hidden;
  height: 500px;
  width: 100%;

  &__arrow-button {
    position: absolute;
    z-index: 2;
    top: 50%;
    transform: translateY(-50%);

    &:nth-child(2) {
      right: 0;
    }
  }

  &-slides {
    position: relative;
    display: flex;
    width: fit-content;
    height: 100%;
  }

  &-pagination {
    @include flex;
    position: absolute;
    z-index: 2;
    bottom: 5%;
    right: 50%;
    transform: translateX(50%);

    &__item {
      transition: transform 100ms linear;
      margin-right: 3px;
      width: 14px;
      height: 8px;
      border: 1px solid $alto;
      cursor: pointer;

      &:last-child {
        margin-right: 0;
      }

      &--active {
        background: $alto;
      }

      &:hover {
        transform: scale(1.1);
      }
    }
  }

  .transition {
    transition: transform ease-in-out 0.01ms;
  }

  .transitionCarousel {
    transition: transform 1.5s ease-in-out;
  }
}

@include media-for(mobile-landscape) {
  .slideshow {
    height: 400px;
  }
}
