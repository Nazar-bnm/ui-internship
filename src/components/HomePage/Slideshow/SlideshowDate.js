import girlInOveralls from '@/assets/img/background/girl-in-overalls.jpg';
import girlOnBoat from '@/assets/img/background/girl-on-boat.jpg';
import girls from '@/assets/img/background/girls.jpg';
import manInCoat from '@/assets/img/background/men-in-coat.jpg';

export const slideshowData = [
  {
    id: 1,
    img: girlInOveralls,
    title: 'new collection',
    description: 'Special autumn edition',
    buttonName: 'discover the editorial'
  },
  {
    id: 2,
    img: girlOnBoat,
    title: 'new collection',
    description: 'a long journey',
    buttonName: 'find a perfect outfit'
  },
  {
    id: 3,
    img: girls,
    title: 'sales',
    description: 'a new look',
    buttonName: 'start shopping now'
  },
  {
    id: 4,
    img: manInCoat,
    title: 'new collection',
    description: 'a new wave',
    buttonName: 'find a perfect outfit'
  }
];
