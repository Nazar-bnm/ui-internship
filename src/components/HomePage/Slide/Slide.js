/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import './Slide.scss';

const slide = ({ bgImage, animation, onAnimationEnd, style, children }) => {
  const bgStyle = {
    backgroundImage: `url(${bgImage})`,
    backgroundRepeat: 'no-repeat',
    backgroundSize: 'cover',
    backgroundPosition: 'center center',
  };

  return (
    <div className={'slide-container'} style={style}>
      <div className={`slide-info ${animation}`} style = {bgStyle}onAnimationEnd={onAnimationEnd}>
        {children}
      </div>
    </div>
  );
};


slide.propTypes = {
  left: PropTypes.bool,
  right: PropTypes.bool,
  clicked: PropTypes.func,
  className: PropTypes.string,
};

slide.defaultProps = {
  left: false,
  right: false,
  clicked: () => {},
  className: '',
};

export default slide;
