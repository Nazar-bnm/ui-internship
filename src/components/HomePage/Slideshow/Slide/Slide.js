import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import './Slide.scss';

const slide = ({ bgImage, animation, onAnimationEnd, style, children }) => {
  const bgStyle = {
    backgroundImage: `url(${bgImage})`,
  };

  const CN = 'slide';

  return (
    <div className={`${CN}-container`} style={style}>
      <div className={cx(`${CN}-wrapper`, [animation])} style = {bgStyle} onAnimationEnd={onAnimationEnd}>
        {children}
      </div>
    </div>
  );
};

slide.propTypes = {
  bgImage: PropTypes.string,
  animation: PropTypes.string,
  onAnimationEnd: PropTypes.func,
  style: PropTypes.object,
  children: PropTypes.node,
};

slide.defaultProps = {
  bgImage: '',
  animation: '',
  onAnimationEnd: () => {},
  style: {},
  children: PropTypes.node,
};

export default slide;
