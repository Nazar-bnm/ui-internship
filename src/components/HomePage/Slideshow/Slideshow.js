import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import ArrowButton from './ArrowButton';
import Slide from './SlideContainer';
import { ANIMATION_NAMES, ARROW_BUTTON_TYPES } from '../../../constants/SlideshowConst';

import './Slideshow.scss';

const {
  CAROUSEL,
  FADE,
  ZOOM_IN,
  ZOOM_OUT
} = ANIMATION_NAMES;
const { LEFT, RIGHT } = ARROW_BUTTON_TYPES;
const CN = 'slideshow';

class Slideshow extends Component {
  constructor(props) {
    super(props);

    this.state = {
      index: 0,
      prevIndex: null,
      autoPlay: true,
      isClicked: false,
      shouldTransition: true,
      isTransform: false
    };

    this.sliderInterval = null;
    this.nextSlide = this.nextSlide.bind(this);
    this.transitionEnd = this.transitionEnd.bind(this);
    this.removeTransform = this.removeTransform.bind(this);
    this.deleteAutoPlay = this.deleteAutoPlay.bind(this);
    this.onMouseLeftHandler = this.onMouseLeftHandler.bind(this);
    this.changeSlideWithButton = this.changeSlideWithButton.bind(this);
  }

  componentDidMount() {
    this.startSlideShow();
  }

  componentDidUpdate(prevProps, { autoPlay: prevStateAutoPlay }) {
    const { autoPlay } = this.state;

    if (autoPlay && autoPlay !== prevStateAutoPlay) {
      this.startSlideShow();
    }
  }

  onMouseLeftHandler() {
    const { isClicked } = this.state;

    !isClicked
      && this.setState({
        autoPlay: true
      });
  }

  isAnimationZoom() {
    const { animation } = this.props;

    return animation === ZOOM_IN || animation === ZOOM_OUT;
  }


  updateIndex(number, newIndex) {
    this.setState(({ index: prevIndex }) => ({
      index: newIndex !== undefined ? newIndex : prevIndex + number,
      shouldTransition: true,
      ...(this.isAnimationZoom()
        ? { isTransform: true, prevIndex }
        : {})
    }));
  }

  nextSlide() {
    const { index } = this.state;
    const { slideData } = this.props;

    index !== slideData.length && this.updateIndex(1);
  }

  prevSlide() {
    const { index } = this.state;
    const indexOfLastCloneSlide = -1;

    index !== indexOfLastCloneSlide && this.updateIndex(-1);
  }

  changeSlideWithButton(type) {
    this.setState({
      isClicked: true
    });

    type === RIGHT && this.nextSlide();
    type === LEFT && this.prevSlide();
  }

  changeSlideWithPagination(e, i) {
    const { index } = this.state;

    this.setState({ isClicked: true });
    i !== index && this.updateIndex('', i);
  }

  deleteAutoPlay() {
    this.setState({
      autoPlay: false
    });

    clearInterval(this.sliderInterval);
  }

  startSlideShow() {
    this.sliderInterval = setInterval(this.nextSlide, 4000);
  }

  transitionEnd() {
    const { index } = this.state;
    const { slideData } = this.props;
    const slideDataLength = slideData.length;
    const indexOfLastCloneSlide = -1;
    const indexOfLastSlide = slideDataLength - 1;

    index === slideDataLength
      && this.setState({
        index: 0,
        shouldTransition: false
      });

    index === indexOfLastCloneSlide
      && this.setState({
        index: indexOfLastSlide,
        shouldTransition: false
      });
  }

  removeTransform() {
    this.setState({
      isTransform: false
    });
  }

  renderPagination() {
    const { index } = this.state;
    const { slideData } = this.props;

    return slideData.map((el, dataIndex) => (
      <div
        key={el.id}
        className={cx(`${CN}-pagination__item`, {
          [`${CN}-pagination__item--active`]: index === dataIndex
        })}
        onClick={(e) => this.changeSlideWithPagination(e, dataIndex)}
      />
    ));
  }

  renderSlides() {
    const { index, prevIndex, isTransform } = this.state;
    const { slideData, animation } = this.props;
    const transformSlide = `translateX(${100 * (index - prevIndex)}%)`;

    const {
      0: firstCloneData,
      length,
      [length - 1]: lastCloneData
    } = slideData;

    const slides = slideData.map((el, i) => (
      <Slide
        key={el.id}
        bgImage={el.img}
        transform={
          prevIndex === i && isTransform ? transformSlide : 'none'
        }
        animation={cx({
          [animation]:
            this.isAnimationZoom() && i === prevIndex,
          [FADE]: animation === FADE && i === index
        })}
        onAnimationEnd={this.removeTransform}
      >
        {el.component}
      </Slide>
    ));

    return [
      <Slide key="lastClone" bgImage={lastCloneData.img}>
        {lastCloneData.component}
      </Slide>,
      ...slides,
      <Slide key="firstClone" bgImage={firstCloneData.img}>
        {firstCloneData.component}
      </Slide>
    ];
  }

  render() {
    const { index, shouldTransition } = this.state;
    const { className, animation } = this.props;
    const slides = this.renderSlides();

    const styleContainer = {
      transform: `translateX(-${(100 / slides.length)
        * (index + 1)}%)`
    };

    return (
      <div
        className={cx(`${CN}`, className)}
        onMouseEnter={this.deleteAutoPlay}
        onMouseLeave={this.onMouseLeftHandler}
      >
        <ArrowButton
          type={LEFT}
          className={`${CN}__arrow-button`}
          onClick={() => this.changeSlideWithButton(LEFT)}
        />
        <ArrowButton
          type={RIGHT}
          className={`${CN}__arrow-button`}
          onClick={() => this.changeSlideWithButton(RIGHT)}
        />
        <div
          className={cx(`${CN}-slides`, {
            transition: shouldTransition,
            transitionCarousel:
              shouldTransition && animation === CAROUSEL
          })}
          style={styleContainer}
          onTransitionEnd={this.transitionEnd}
        >
          {slides}
        </div>
        <div className={`${CN}-pagination`}>{this.renderPagination()}</div>
      </div>
    );
  }
}

const dataItemProps = PropTypes.shape({
  id: PropTypes.oneOfType([PropTypes.string,
    PropTypes.number]),
  img: PropTypes.string,
  component: PropTypes.element
});

Slideshow.propTypes = {
  className: PropTypes.string,
  animation: PropTypes.oneOf([CAROUSEL, FADE, ZOOM_IN, ZOOM_OUT]),
  slideData: PropTypes.arrayOf(dataItemProps)
};

Slideshow.defaultProps = {
  className: '',
  animation: CAROUSEL,
  slideData: []
};

export default Slideshow;
