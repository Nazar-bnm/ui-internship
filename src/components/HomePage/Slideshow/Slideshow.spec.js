import React from 'react';
import { shallow } from 'enzyme';

import Slideshow from './Slideshow';
import Slide from '../Slide';
import { slideshowData } from './SlideshowDate';
import { ANIMATION_NAMES } from '@/constants/SlideshowConst';

describe('<Slideshow />', () => {
  const CN = 'slideshow-slides';
  const { CAROUSEL } = ANIMATION_NAMES;
  const slideDataLength = slideshowData.length;

  const slidesData = slideshowData.map((el) => {
    return {
      id: el.id,
      img: el.img,
      component: (
        <Slide
          title={el.title}
          description={el.description}
          buttonName={el.buttonName}
        />
      ),
    };
  });

  let wrapper;
  let index;

  beforeEach(() => {
    wrapper = shallow(<Slideshow animation={CAROUSEL} slideData={slidesData} />);
  });

  it('should render component <Slideshow />', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render pagination items that is equal to slideData length', () => {
    expect(wrapper.find('.slideshow-pagination__item')).toHaveLength(slideDataLength);
  });

  it('expect index of the state to be index of clicked pagination item ', () => {
    index = 0;
    wrapper.find('.slideshow-pagination__item').at(index).simulate('click');
    expect(wrapper.state().index).toBe(index);
  });

  it('expect the clicked pagination item has active class but not the rest', () => {
    wrapper.find('.slideshow-pagination__item').forEach((element) => {
      element.simulate('click');
    });

    wrapper.find('.slideshow-pagination__item').forEach((element, index) => {
      if (index != slideDataLength - 1) {
        expect(element.hasClass('slideshow-pagination__item--active')).toBe(false);
      } else {
        expect(element.hasClass('slideshow-pagination__item--active')).toBe(true);
      }
    });
  });

  it('should pause autoPlay when mouse enter Slideshow component', () => {
    wrapper.simulate('mouseenter');
    expect(wrapper.state().autoPlay).toBe(false);
  });

  it('should start autoPlay when mouse left Slideshow component and none of buttons was clicked', () => {
    wrapper.simulate('mouseenter');
    wrapper.simulate('mouseleave');
    expect(wrapper.state().autoPlay).toBe(true);
    wrapper.simulate('mouseenter');
    wrapper.find('.slideshow__arrow-button').first().simulate('click');
    wrapper.simulate('mouseleave');
    expect(wrapper.state().autoPlay).toBe(false);
  });

  it('should increase index of state by 1 when click right arrow,'+
  'but after slideDataLength clicks and when transition ends index should be again equal to 0', () => {
    wrapper.find('[type="right"]').simulate('click');
    wrapper.find(`.${CN}`).simulate('transitionend');
    expect(wrapper.state().index).toBe(1);
    wrapper.find('[type="right"]').simulate('click');
    expect(wrapper.state().index).toBe(2);
    wrapper.find('[type="right"]').simulate('click');
    wrapper.find('[type="right"]').simulate('click');
    wrapper.find('[type="right"]').simulate('click');
    wrapper.find(`.${CN}`).simulate('transitionend');
    expect(wrapper.state().index).toBe(0);
  });

  it('should decrease index of state by 1 when click right arrow'+
  'but after first click and when transition ends index should be equal to slideDataLength - 1',
  () => {
    wrapper.find('[type="left"]').simulate('click');
    expect(wrapper.state().index).toBe(-1);
    wrapper.find(`.${CN}`).simulate('transitionend');
    expect(wrapper.state().index).toBe(slideDataLength - 1);
    wrapper.find('[type="left"]').simulate('click');
    expect(wrapper.state().index).toBe(slideDataLength - 2);
  });

  it('should pause autoPlay when mouse enter Slideshow component', () => {
    wrapper.simulate('mouseenter');
    expect(wrapper.state().autoPlay).toBe(false);
  });

  it('should set transform when index changes and remove when animation ends', () => {
    index = 3;
    wrapper = shallow(<Slideshow animation="zoom-in" slideData={slideshowData} />);
    wrapper.find('.slideshow-pagination__item').at(index).simulate('click');
    expect(wrapper.state().isTransform).toBe(true);
    const prevIndex = wrapper.state().prevIndex;
    wrapper.find('[animation="zoom-in"]').at(prevIndex).simulate('animationend');
    expect(wrapper.state().isTransform).toBe(false);
    index = 2;
    wrapper.find('.slideshow-pagination__item').at(index).simulate('click');
    expect(wrapper.state().isTransform).toBe(true);
  });
});
