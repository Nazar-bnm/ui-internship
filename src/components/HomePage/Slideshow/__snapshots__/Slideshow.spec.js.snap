// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slideshow /> should render component <Slideshow /> 1`] = `
<div
  className="slideshow"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <arrowButton
    className="slideshow__arrow-button"
    onClick={[Function]}
    type="left"
  />
  <arrowButton
    className="slideshow__arrow-button"
    onClick={[Function]}
    type="right"
  />
  <div
    className="slideshow-slides transition transitionCarousel"
    onTransitionEnd={[Function]}
    style={
      Object {
        "transform": "translateX(-16.666666666666668%)",
      }
    }
  >
    <SlideContainer
      animation=""
      bgImage={
        Object {
          "process": [Function],
        }
      }
      key="lastClone"
      onAnimationEnd={[Function]}
      style={Object {}}
    >
      <Component />
    </SlideContainer>
    <SlideContainer
      animation=""
      bgImage={
        Object {
          "process": [Function],
        }
      }
      key="1"
      onAnimationEnd={[Function]}
      style={Object {}}
    >
       
    </SlideContainer>
    <SlideContainer
      animation=""
      bgImage={
        Object {
          "process": [Function],
        }
      }
      key="2"
      onAnimationEnd={[Function]}
      style={Object {}}
    >
       
    </SlideContainer>
    <SlideContainer
      animation=""
      bgImage={
        Object {
          "process": [Function],
        }
      }
      key="3"
      onAnimationEnd={[Function]}
      style={Object {}}
    >
       
    </SlideContainer>
    <SlideContainer
      animation=""
      bgImage={
        Object {
          "process": [Function],
        }
      }
      key="4"
      onAnimationEnd={[Function]}
      style={Object {}}
    >
       
    </SlideContainer>
    <SlideContainer
      animation=""
      bgImage={
        Object {
          "process": [Function],
        }
      }
      key="firstClone"
      onAnimationEnd={[Function]}
      style={Object {}}
    >
      <Component />
    </SlideContainer>
  </div>
  <div
    className="slideshow-pagination"
  >
    <div
      className="slideshow-pagination__item slideshow-pagination__item--active"
      key="0"
      onClick={[Function]}
    />
    <div
      className="slideshow-pagination__item"
      key="1"
      onClick={[Function]}
    />
    <div
      className="slideshow-pagination__item"
      key="2"
      onClick={[Function]}
    />
    <div
      className="slideshow-pagination__item"
      key="3"
      onClick={[Function]}
    />
  </div>
</div>
`;
